File: src/main/java/utilities/YamlParser.java
package utilities;

import graph_components.Property;
import org.yaml.snakeyaml.Yaml;

import java.util.*;

public class YamlParser {
    private final Yaml yaml = new Yaml();

    public List<Property> parseProperties(String yamlPropertiesString) {
        Map<String, Object> values = yaml.load(yamlPropertiesString);
        List<Property> propertiesFromString = new ArrayList<>();

        values.forEach((key, value) -> {
            if (value instanceof String) {
                propertiesFromString.add(new Property(key, String.valueOf(value)));
            } else if (value instanceof LinkedHashMap) {
                propertiesFromString.add(fromMap(key, value));
            }
        });


        return propertiesFromString;
    }

    private static Property fromMap(String propertyKey, Object map) {

        LinkedHashMap<String, Object> prop = (LinkedHashMap<String, Object>) map;

        String key = prop.keySet().iterator().next();

        if (prop.get(key) instanceof List) {
            return new Property(propertyKey, key, (List<Object>) prop.get(key));
        }

        throw new IllegalArgumentException("Invalid format");
    }

}

File: src/main/kotlin/utilities/YamlParser.kt
package utilities

import graph_components.Property
import org.yaml.snakeyaml.Yaml
import java.util.*

class YamlParser {
    val yaml = Yaml()

    fun parseProperties(yamlPropertiesString: String): List<Property> {
        val values = yaml.load<Map<String, Any>>(yamlPropertiesString)
        val properties = arrayListOf<Property>()
        for ((key, value) in values) {
            if (value is String) {
                properties.add(Property(key, value))
            } else {
                properties.add(fromMap(key, value))
            }
        }

        return properties
    }


    private fun fromMap(propertyKey: String, map: Any): Property {

        @Suppress("UNCHECKED_CAST") val prop = map as LinkedHashMap<String, Any>

        val key = prop.keys.iterator().next()

        if (prop[key] is List<*>) {
            return (Property(propertyKey, key, (prop[key] as List<*>).filterIsInstance<Any>()))
        }

        throw IllegalArgumentException("Invalid format")
    }
}
